ballerina/xmldata.bal

```diff
@@ -43,53 +43,63 @@ public type JsonOptions record {
+                            check getAttributesMap(jsonValue, allNamespaces, options = options));
     } else {
         map<json>|error jMap = jsonValue.ensureType();
         if jMap is map<json> {
             if jMap.length() == 0 {
                 return xml ``;
             }
             json value = jMap.toArray()[0];
+            addNamespaces(allNamespaces, check getNamespacesMap(value, {}, options));
             if value is json[] {
-                return getElement("root", check traverseNode(value, {}, options, jMap.keys()[0]),
-                                check getAttributesMap(value, options = options));
+                return getElement("root", check traverseNode(value, allNamespaces, {}, options, jMap.keys()[0]),
+                                allNamespaces, check getAttributesMap(value, allNamespaces, options = options));
             } else {
                 string key = jMap.keys()[0];
                 if key == CONTENT {
                     return xml:createText(value.toString());
                 }
-                return getElement(jMap.keys()[0], check traverseNode(value, {}, options),
-                                check getAttributesMap(value, options = options));
+                addNamespaces(allNamespaces, check getNamespacesMap(value, {}, options));
+                return getElement(jMap.keys()[0], check traverseNode(value, allNamespaces, {}, options), allNamespaces,
+                                check getAttributesMap(value, allNamespaces, options = options));
             }
         }
         if jsonValue !is null {
-           return xml:createText(jsonValue.toString());
+            return xml:createText(jsonValue.toString());
         } else {
-           return xml ``;
+            return xml ``;
         }
     }
 }
 
-isolated function traverseNode(json jNode, map<string> parentNamespaces, JsonOptions options = {},
-                                string? key = ()) returns xml|Error {
+isolated function traverseNode(json jNode, map<string> allNamespaces, map<string> parentNamespaces,
+                               JsonOptions options = {}, string? key = ()) returns xml|Error {
     string arrayEntryTag = options.arrayEntryTag == "" ? "item" : options.arrayEntryTag;
     string attributePrefix = options.attributePrefix == "" ? "@" : options.attributePrefix;
+    map<string> namespacesOfElem = {};
     xml xNode = xml ``;
     if jNode is map<json> {
         foreach [string, json] [k, v] in jNode.entries() {
             if !k.startsWith(attributePrefix) {
                 if k == CONTENT {
                     xNode += xml:createText(v.toString());
                 } else if v is json[] {
-                    xml node = check traverseNode(v, check getNamespacesMap(v, parentNamespaces, options), options, k);
+                    namespacesOfElem = check getNamespacesMap(v, parentNamespaces, options);
+                    addNamespaces(allNamespaces, namespacesOfElem);
```

Comment:
Can we move this out as we call the same in both clauses?

Reply:
Updated.